### PR TITLE
FEAT: help recoded (issue #2123, issue #2127, issue #2128)

### PR DESCRIPTION
**Pull request by Oldes**
This modification provides more Rebol3 like help output.

### CODE REVIEW

**Comment by qtxie** [comment]
Looks nice! 👍

**Comment by Oldes** [comment]
I removed datatype definitions as these were taken from Rebol source. Could be added back if someone will write what should be printed when one uses: `help datatype!`.

The `help` is again re-coded to reduce it's size and efficiency. I removed `fetch-help` function and instead added refinement `into` in the main `help` function. With this refinement it is possible to get help result into string instead of being printed in the console.

Block listed in the object is now showing partial content instead of block length information as it is in Rebol.

**Comment by Oldes** [comment]
I modified the output of the help function... main difference is, that it crop description lines in object listing according width of the console (gui version only so far) and it correctly handle multiple strings in function descriptions... so:
```
my-function: func [
	"This is the best function"
	"It can handle almost everything!"
	arg        "It takes this argument"
	           "Don't know, why I write this?"
	/path      "Also path can be used"
	 p [path!] "Use this to specify location or whatever"
	return: [any-string!]
][]

? my-function
```

will show:
```
USAGE:
    MY-FUNCTION arg

DESCRIPTION:
    This is the best function. 
    It can handle almost everything! 
    MY-FUNCTION is a function! value.
    Returns value of type: [any-string!]

ARGUMENTS:
    arg        [any-type!] It takes this argument. Don't know, why I write this?

REFINEMENTS:
    /path      <- Also path can be used
       p       [path!]    Use this to specify location or whatever.
```

**Comment by Oldes** [comment]
Just for comparison.. if used without the `return:` part, which is not supported in Rebol, this would be output in Rebol 2:
```
USAGE:
    MY-FUNCTION arg /path p

DESCRIPTION:
     This is the best function
     MY-FUNCTION is a function value.

ARGUMENTS:
     arg -- It takes this argument (Type: any)

REFINEMENTS:
     /path -- Also path can be used
         p -- Use this to specify location or whatever (Type: path)
```
and this in Rebol 3:
```
USAGE:
        MY-FUNCTION arg /path p

DESCRIPTION:
        This is the best function
        MY-FUNCTION is a function value.

ARGUMENTS:
        arg -- Don't know, why I write this?

REFINEMENTS:
        /path -- Also path can be used
                p -- Use this to specify location or whatever (path!)
```

File: environment/console/help.red [hunk 1]
```diff
@@ -59,226 +36,305 @@ Other useful functions:
...
-					]
-					any [
-						/local to end
-						| set w [word! | lit-word! | get-word!] (prin-out out " " prin-out out w)
-						| set w refinement! (prin-out out " " prin-out out mold w)
-						| skip
-					]
+
+		catch [case/all [
+			if unset? :word [									;-- HELP with no arguments
```
Comment (dockimbel) [inline environment/console/help.red]: Seems like that `if` is a leftover of older code, it should be removed.

File: environment/console/help.red [hunk 2]
```diff
@@ -59,226 +36,305 @@ Other useful functions:
...
+				buffer: insert buffer HELP-USAGE
+				throw true
+			]
+
+			if word? :word [
```
Comment (dockimbel) [inline environment/console/help.red]: Same remark for this `if` as for previous one.

File: environment/console/win32.reds [hunk 3]
```diff
@@ -197,37 +199,48 @@ fd-read: func [
 		key 	[key-event!]
 		n	 	[integer!]
 		keycode [integer!]
+		size    [red-pair!]
 ][
 	n: 0
 	while [true] [
```
Comment (dockimbel) [inline environment/console/win32.reds]: `while [true]` => `forever`

**Comment by dockimbel** [comment]
> it correctly handle multiple strings in function descriptions

Multiple strings are allowed, but they are not meant for splitting the same docstring into multiple pieces (that would be a waste), it is meant to allow for extra informations to be encoded in the future in additional strings (we could then define a convention for the meaning of the 1st, 2nd, ... docstrings). It's an option for future use.

File: environment/console/win32.reds (hunk 3 above)
Comment (Oldes) [inline environment/console/win32.reds]: The `while [true]` part is from the original... also... this part of code should not be part of this pull request probably.. it leaked there from the `system/console/limit` changes here: https://github.com/red/red/pull/2514

Comment (Oldes) [inline environment/console/win32.reds]: Although the help already depends on the value from the `system/console/size`

**Comment by Oldes** [comment]
> Multiple strings are allowed, but they are not meant for splitting the same docstring into multiple pieces (that would be a waste). It is meant to allow for extra information to be encoded in additional strings (we could then define a convention for the meaning of the 1st, 2nd, ... docstrings). It's an option for future use.

So what is expected output for such a situation? Above are 3 versions.. there can be also 4. version where the multiple doc strings are printed on own lines.

**Comment by Oldes** [comment]
@dockimbel I was also thinking about different output for this situation:
```
quit: func [
	"Stops evaluation and exits the program"
	/return status	[integer!] "Return an exit status"
][
	quit-return any [status 0]
]
```
It would now show:
```
REFINEMENTS:
    /return
       status  [integer!] Return an exit status.
```
Maybe it would be better to detect such a situation and display:
```
REFINEMENTS:
    /return    <- Return an exit status.
       status  [integer!] 
```
Or maybe the change should be done in the function source... there are many undocumented or ill documented functions.

**Comment by Oldes** [comment]
Btw... there is a bug in above situation as `/return` is detected as a `returning` result . I will fix it.

**Comment by dockimbel** [comment]
As @greggirwin version was merged, it would be good to port the fixes provided by this PR for the mentioned tickets above to the current version in master. @Oldes I'm closing this PR now, feel free to submit a new one with extra fixes/features required for closing those tickets.

**Comment by Oldes** [comment]
It looks that the most important difference - the real console size - was incorporated by @greggirwin already, so I'm probably fine.

In the new version I don't like the pipe character, especially when there is no following content and long name, like here:

image

**Comment by Oldes** [comment]
@dockimbel @greggirwin from the mentioned issues, only issue #2127 is not fixed with current version.

**Comment by greggirwin** [comment]
I'll review comments here and in other tickets. Changing the | is easy, if we just want a different char. Also easy to leave it out if there is no value. 

The two things on the top of my list are to fix alignment/padding issues and add the ability to get help output as a string, if we still want that feature. On padding, the big question is how important it is to have a fully polished output, which means we'd have to find the longest name in any list and format based on that. Not hard, just overhead.

**Comment by greggirwin** [comment]
One of the downsides with padding to the longest name, especially with system/words, is that one really long name pushes everything way out.

**Comment by dockimbel** [comment]
@greggirwin In such cases, I usually limit the padding to a fixed size, truncate the long names and add a "..." sequence at tail.

**Comment by PeterWAWood** [comment]
You could make use of Red's Unicode pedigree and save a couple of characters by using …  - #"^(2026)"

**Comment by dockimbel** [comment]
@PeterWAWood Good suggestion, though a special case would need to be made for the CLI console on Windows, as such character would not display properly.

**Comment by greggirwin** [comment]
I'll see how truncating names looks. The problem I forsee is that asking for help and getting only most of name may not seem all that helpful.

**Comment by greggirwin** [comment]
I considered using the ellipsis char in the other truncation scenario, but didn't, for the reason Doc cited.

**Comment by qtxie** [comment]
@greggirwin 
> get help output as a string

Yes, we need it as the VSCode plugin use it when showing the help window.

**Comment by greggirwin** [comment]
I'll make that a priority then Qingtian. Does this also affect what we want, WRT #2127?

**Comment by greggirwin** [comment]
Oldes, one of the things I like about `|` is that it makes the output table-like, though I agree alignment is an issue no matter what. Something like `==` doesn't look as nice to me. `:` is nice, but then there is no hope of making the output loadable, if we wanted to do that. I have added back buffered output here, and experimenting made me think about how we might use that capability.

**Comment by greggirwin** [comment]
I'm adding padding, so object words align nicely, but a ripple effect for `system/words` is that the first column may be wider than our old fixed-width model. That means the console size check can't be static anymore, but will be dependent on the current output (longest word). The console nicely rewraps as you resize, but any new output will have the same issue. And it affects long doc/def strings for all words in the object. One long word, created by any user, will mess up `system/words` output, unless we truncate words. In which case we're back to whether that's acceptable.

**Comment by greggirwin** [comment]
I'm not going to add extra padding, based on longest word, to datatype help right now, because it iterates over `system/words` and would have to do a second pass to do it. It's a single pass op right now, emitting as it finds matches.
